Adding PyRe to cyclus/recycle

gtw2 opened the pull request:
This PR is quite large... it includes the entirety of PyRe - the base process class and each subprocess, a CUSUM detection class, and the diverter class. 

Unit tests are also included so the recycle CI will stop failing due to old cycamore tests.

This has also been reviewed before being PR'd into arfc/recycle.

This will close issues #14 #13 #12 #11 #9

bam241 commented on circle.yml:
```diff
@@ -4,8 +4,8 @@ jobs:
     build:
         docker:
-            - image: cyclus/cyclus:latest
-        working_directory: ~/recyle
```
I don't think this should not depend on `cycamore`....
maybe you can use the sink/source that are provided by `cyclus` ?

bam241 commented on circle.yml:
```diff
@@ -57,14 +57,14 @@ jobs:
 
                 
     # Update docker container
     deploy: # Cycamore -> Cycamore:latest
```
`cycamore` -> `recycle`

bam241 commented on circle.yml:
```diff
@@ -77,15 +77,15 @@ jobs:
-                command: docker push cyclus/recyle:latest # push to docker depot
+                command: docker push cyclus/recycle:latest # push to docker depot
 
     deploy_stable: # Cycamore:stable
```
`cycamore` -> `recycle`

bam241 commented on tests/input/dynamic_capacitated.xml:
```diff
@@ -8,10 +8,10 @@
+    <spec><lib>cycamore</lib><name>Source</name></spec>
     <spec><lib>agents</lib><name>NullRegion</name></spec>
-    <spec><lib>recyle</lib><name>DeployInst</name></spec>
+    <spec><lib>cycamore</lib><name>DeployInst</name></spec>
```
This file should not be part of `recycle`...

bam241 commented on tests/input/growth.xml:
```diff
@@ -9,19 +9,19 @@
 
   <archetypes>
```
same here... not part of `recycle`

bam241 commented on src/cusum.h:
```diff
@@ -0,0 +1,37 @@
+
+bool divert;
+
+public:
```
google style guide for c++ (the one we follow in the cyclus community) suggest:
`A class definition should usually start with a public: section, followed by protected:, then private:. Omit sections that would be empty.`

bam241 commented on src/diverter.cc:
```diff
@@ -0,0 +1,91 @@
+#include "diverter.h"
+
+using cyclus::Material;
+using cyclus::Composition;
```
not used

bam241 commented on src/diverter.cc:
```diff
@@ -0,0 +1,91 @@
+
+using cyclus::Material;
+using cyclus::Composition;
+using cyclus::CompMap;
```
not used

bam241 commented on src/diverter.cc:
```diff
@@ -0,0 +1,91 @@
+using cyclus::Material;
+using cyclus::Composition;
+using cyclus::CompMap;
+using cyclus::Context;
```
not used

bam241 commented on src/diverter.cc:
```diff
@@ -0,0 +1,91 @@
+#include "cyclus.h"
+#include "pyre.h"
+#include "process.h"
+#include "diverter.h"
```
Google Style suggested iclude order:
`Use standard order for readability and to avoid hidden dependencies: Related header, C library, C++ library, other libraries' .h, your project's .h.`

bam241 commented on src/diverter.cc:
```diff
@@ -0,0 +1,91 @@
+#include "cyclus.h"
+#include "pyre.h"
+#include "process.h"
```
this is included twice, it is already in the header... same for `cyclus.h`

bam241 commented on src/diverter.h:
```diff
@@ -0,0 +1,74 @@
+std::map<std::string, Material::Ptr> DivertStream(std::map<std::string, Material::Ptr> sepstreams);
+
+};
+}
```
we usually comment the end of the namespace...

bam241 commented on src/process.cc:
```diff
@@ -0,0 +1,126 @@
+using cyclus::KeyError;
+using cyclus::ValueError;
+using cyclus::Request;
+using cyclus::CompMap;
```
to you really need all of them ?

bam241 commented on src/process.cc:
```diff
@@ -0,0 +1,126 @@
+#include "cyclus.h"
+#include "process.h"
```
include order

bam241 commented on src/process.h:
```diff
@@ -0,0 +1,85 @@
+#define RECYCLE_SRC_PROCESS_H_
+
+#include "cyclus.h"
+#include <vector>
```
include order, cyclus included twice...

bam241 commented on src/process.h:
```diff
@@ -0,0 +1,85 @@
+  double volume();
+
+};
+}
```
end of namespace comment

bam241 commented on src/pyre_reduction.h:
```diff
@@ -0,0 +1,55 @@
+
+public:
+
+Reduct();
```
the indentation is not correct I believe

bam241 commented on src/pyre_refining.cc:
```diff
@@ -0,0 +1,199 @@
+
+using boost::math::tools::bisect;
+
+namespace recycle {
```
I am not sure this should be in the recycle namespace maybe a pyro one ?

bam241 commented on src/pyre_reduction.cc:
```diff
@@ -0,0 +1,71 @@
+using cyclus::Request;
+using cyclus::CompMap;
+
+namespace recycle {
```
shall this be in the `recycle` namesapce ?

bam241 commented on src/pyre_refining.h:
```diff
@@ -0,0 +1,146 @@
+  double p0_;
+  double p1_;
+  double p2_;
+  double p3_;
```
why not a vector ?

bam241 commented on src/pyre_volox.h:
```diff
@@ -0,0 +1,59 @@
+double th0;
+double th1;
+double th2;
+double th3;
```
`vector<double>`

bam241 commented on src/pyre_winning.cc:
```diff
@@ -0,0 +1,77 @@
+                          double c4 = -19.7493
+) {
+  return c0*pow(current(), 4) + c1*pow(current(), 3) + c2*pow(current(), 2)
+          + c3*current() + c4;
```
why this is not using the member variable ?

bam241 replied:
same for all the other methods...

bam241 replied:
and default value should be set in the header, I believe

gtw2 replied:
The reason for using a getter function for current/other variables is because we are keeping track of the time series and the function pulls it off the back of the vector - I believe this was a suggestion by Katy

bam241 replied:
I am not sure to understand what you are saying, but my point is:
`c0`,`c1`, `c2`, `c3`,` c4` are member variable of the class.

as you name them the same way (and you give them default values), the member variable are never not use.
The question is then, is the class required ? or do you just need a set of method to compute the values you need...
also the default values should be in the header not in the `cc` file, so people can know them when looking at the header (which is installed, the cc in packed in the lib and are not necessary easily accessible)

bam241 replied:
does it make sense ?

gtw2 replied:
I think I get what you mean, the coefficients right now will always be their defaults - there currently is no way to change them in the input so they don't need to be a part of the class. Which the only reason they are in the class is for the diverter implementation as seen in the refiner (currently the only place it is put to use)

gtw2 replied:
If I get what you're suggesting it's to have these methods in a different file?

bam241 replied:
it depends on their purpose...

for me there is 2 options there:
- keep the method in the class but use the member variables, (if you don't use the method elsewhere it might be fine)
- have a collection of method describing the physics that you will use, and use them in the class:

something like:
in `physics.h`
```
namespace physics {
double Coulombic(double c0=xx, double c1=yx, ...);
}
```

in `physics.cc`
```
namespace physics {
double Coulombic(double c0, double c1, ...) {
[ ..]

}
}
```

in `pyro_winning.h`
```
double Coulombic();
```
in `pyro_winning.cc`
```
double Winning::Coulombic() {
  return physics::Coulombic(c0, c1, ...);
}

bam241 replied:
the choice between 1 and 2 will depends on how often and in how many place you used those methods, if it is only one place, you probably only need the first solution.

Another the question you need to answer is: would you have to call the Coulombic function of a pyro_wining element using coefficient different than the one stored in the class member...

gtw2 replied:
Ahhh okay I do understand, I think that physics approach would work and be clean - it would also remove the need for the current situation in `pyre_refining.cc`:
`double Refine::Thermal() {
  return Thermal(temp(),0);
}

double Refine::Thermal(double tmp,
                       double new_eff
) {
  return t0()*pow(tmp,3) + t1()*pow(tmp,2) + t2()*tmp + t3() - new_eff;
}`

gtw2 replied:
The coefficients will differ depending on your input facility. For example the values used now are for kaeri's pride facility whereas this function would differ for INL or ANL's

bam241 commented on src/pyre_winning.h:
```diff
@@ -0,0 +1,61 @@
+#include "process.h"
+#include "recycle_version.h"
+
+class Process;
```
probably not required as you included `process.h`

bam241 commented on src/pyre_refining.cc:
```diff
@@ -0,0 +1,199 @@
+}
+
+double Refine::PressureEff() {
+  return p0()*pow(pressure(), 3) + p1()*pow(pressure(), 2) + p2()*pressure() + p3();
```
why not calling the next method `PressureEff(double, double)` passing the member variable (so you have only one implementation of the function -> so only one to maintain :)

bam241 commented on src/pyre_refining.cc:
```diff
@@ -0,0 +1,199 @@
+    }
+  }
+  return agi;
+}
```
same here...

gtw2 commented:
Thank you for the detailed review @bam241! I'll get to work addressing your comments, as for the CI PR you mentioned. I imagine that PR would have to be merged first since the tests wouldn't pass with the previous CI, right?

bam241 commented:
yes i guess, but it should be fairly simple, as you already have most of the change in the PR `#20 from gtw2/dev`

bam241 commented:
@gtw2 you don't have to remove those change form you PR... just issue a new PR with those change.

once the CI PR is merged, you can just merge master and your branch and it should be good.
Does it make sense ?

gtw2 commented:
Ah yes, that is much simpler... I'm afraid I've overthought it...

bam241 commented:
Nice we got the CI part clean :)

bam241 commented:
@gtw2 ping me when this is ready for an other look :)

gtw2 commented:
Will do! Those cmake fixes took most of my time until now, (haven't done much with cmake so took a while to figure out) will be cleaning up the style/variables you've commented on soon

bam241 commented on CMakeLists.txt:
```diff
@@ -257,6 +257,7 @@ IF(NOT CYCLUS_DOC_ONLY)
 
     # include all the directories we just found
     INCLUDE_DIRECTORIES(${RECYCLE_INCLUDE_DIRS})
+    INCLUDE_DIRECTORIES(src src/toolkit src/toolkit/pyroprocessing)
```
I don't think it is necessary to add `src/toolkit` and `src/toolkit/pyroprocessing`:
as you added `toolkit` in `src/CMakelists.txt`:
the main `CMakeLists.txt` will include `src` which will trigger the processing of `src/CMakeLists.txt` which includes `src/toolkit` so it will trigger the processing of `src/toolkit/CMakeLists.txt` and so on...

gtw2 replied:
That makes sense, I've just tested that and to get the includes to work properly I have to add the path: i.e. in diverter.h `#include "pyroprocessing/process.h"`. Is this better convention?

gtw2 replied:
Actually, I run into a build issue with pyre_tests when doing this. The test file cannot find the pyroprocessing directory

`In file included from /home/gtw2/recycle/build/recycle/pyre_tests.cc:1:0:
/home/gtw2/recycle/src/pyre.h:6:39: fatal error: pyroprocessing/pyre_volox.h: No such file or directory
 #include "pyroprocessing/pyre_volox.h"
`

bam241 commented:
Probably need to #include “toolkit/pyroprocessing/pyre_volox.h”…


> On Aug 5, 2019, at 16:26, Greg Westphal <notifications@github.com> wrote:
> 
> @gtw2 commented on this pull request.
> 
> In CMakeLists.txt <https://github.com/cyclus/recycle/pull/16#discussion_r310799034>:
> 
> > @@ -257,6 +257,7 @@ IF(NOT CYCLUS_DOC_ONLY)
>  
>      # include all the directories we just found
>      INCLUDE_DIRECTORIES(${RECYCLE_INCLUDE_DIRS})
> +    INCLUDE_DIRECTORIES(src src/toolkit src/toolkit/pyroprocessing)
> Actually, I run into a build issue with pyre_tests when doing this. The test file cannot find the pyroprocessing directory
> 
> In file included from /home/gtw2/recycle/build/recycle/pyre_tests.cc:1:0: /home/gtw2/recycle/src/pyre.h:6:39: fatal error: pyroprocessing/pyre_volox.h: No such file or directory #include "pyroprocessing/pyre_volox.h"
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cyclus/recycle/pull/16?email_source=notifications&email_token=ADTRSOQBHHCAST354AFLQCDQDCLI3A5CNFSM4IITHOUKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCATVTEI#discussion_r310799034>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ADTRSOXRKOHJIFZ5VZHFBJTQDCLI3ANCNFSM4IITHOUA>.
>

gtw2 commented:
Yup, that was it.

gtw2 commented:
This most recent commit/PR is more in line with what you mentioned for Option 1 I think. I realized that each process can have different values for those functions - so I've kept them in the class for now. 

Let me know if I'm still off from what you wanted though.

gtw2 commented:
Have you had a chance to look at this @bam241?

bam241 commented:
sorry I just came back from vacation. I'll look at it asap !

gtw2 commented:
Thanks!

bam241 commented on src/toolkit/diverter.cc:
```diff
@@ -0,0 +1,85 @@
+Diverter::Diverter() {}
+
+Diverter::Diverter(std::pair<std::string,std::string> location, 
+    int frequency = 1E299, double quantity = 0.01, int divert_number = 1, std::string type_ = "operator") {
```
I would prefer the default value to be in the header file not the cpp

bam241 replied:
this is usually how we did this in cycamore/cyclus

bam241 commented on src/toolkit/pyroprocessing/pyre_refining.h:
```diff
@@ -0,0 +1,145 @@
+  double a0_;
+  double a1_;
+  double a2_;
+  double a3_;
```
maybe vector would be better ?

bam241 commented on src/toolkit/pyroprocessing/pyre_refining.cc:
```diff
@@ -0,0 +1,185 @@
+  a0_ = 0.032;
+  a1_ = 0.72;
+  a2_ = 0.0338396;
+  a3_ = 0.83667;
```
You should avoid Magic numbers.
maybe they should be set in the constructor,
and we need  a reference about them... to explain where they come from...

gtw2 replied:
Good point, I'll add comments for the papers used to generate/get these numbers

bam241 commented on src/toolkit/pyroprocessing/pyre_refining.cc:
```diff
@@ -0,0 +1,185 @@
+
+struct Refine::TerminationCondition {
+    bool operator() (double min, double max)  {
+        return abs(min - max) <= 0.000001;
```
why `0.00001` and not `0.0000001`

also maybe `1e-5` would be more readable...

bam241 commented on src/toolkit/pyroprocessing/pyre_volox.cc:
```diff
@@ -0,0 +1,60 @@
+  ti.push_back(-1.696);
+  r.push_back(0.12435);
+  r.push_back(0.7985);
+}
```
same here where those magic number come from ?
to what do they corresponds ?

bam241 replied:
Something bother me with the way you set those coefficient... would it be possible to have default value set in the header, where the reference of them is ?
something like:
`vector<double> th = {4.369E-9, ... , 0.2903};`

bam241 commented on src/toolkit/pyroprocessing/pyre_winning.cc:
```diff
@@ -0,0 +1,62 @@
+                 double new_time = 1, 
+                 double new_flowrate = 3, 
+                 double new_volume = 1
+            ) 
```
default value in the header ?

bam241 replied:
with you actual implementation how do you know which Winning() constructor is actually called when calling Winning::Winning() ? (as the second one has 4 optional arguments ) (is that even compiling ?!??)

gtw2 replied:
I'm not sure what you mean, when it's constructed in pyre it is called using the 4 inputs so it is using the overloaded constructor (I do need to document the default constructor though).

The scenarios I've run have all received those 4 parameters so I'm fairly certain this is what happens. 

That being said though I can see some of the confusion as these default values are also given in the pragmas in pyre.h so are they really needed here? (or in the header as you suggest)

bam241 replied:
This is not `pyre` class, this might be use by someone else in an other facility so I believe you should have the default value regardless of your usage in `pyre`

About the constructor,  I mean you have 2 constructor methods with the same name one without argument and one with 4 optional arguments (as they have default values), when you call the the constructor without arguments which one are you calling, the first one that set the value to `0` or the one using the default values ?

gtw2 commented on src/toolkit/pyroprocessing/pyre_winning.h:
```diff
@@ -0,0 +1,54 @@
+  /// @brief Material separation as a function of flowrate
+  /// @return a value between 0 and 1 relating to separation efficiency
+  double RateEff();
+
```
doc

gtw2 replied:
This comment was about missing documentation for the Throughput function shown below, and has been added.

gtw2 commented:
Sorry for the delay in getting these fixes to you @bam241 - but I've added a couple tests for each pyre_process as you suggested. These test that the classes function at max and min parameters. 

The concern you had about the optional parameters in winning were addressed by removing the default value from the overloaded constructor and adding them to the default one (previous the default constructor used all zeros).

recycle/src was added to RECYCLE_INCLUDE_DIRS for a cleaner INCLUDE_DIRECTORIES call.

Magic numbers have had their references included, and coefficients that should have been in a vector are now correctly vectored.

bam241 commented on src/toolkit/pyroprocessing/pyre_refining.h:
```diff
@@ -0,0 +1,139 @@
+  double a2();
+
+  /// @return the value for rotation efficiency coefficient 3
+  double a3();
```
I don't understand where those methods are implemented..

gtw2 replied:
These are deprecated since changing to vectors, I've removed them

bam241 commented on src/toolkit/pyroprocessing/pyre_winning.cc:
```diff
@@ -0,0 +1,62 @@
+  ti.push_back(0.2903);
+  ti.push_back(-1.696);
+  r.push_back(0.12435);
+  r.push_back(0.7985);
```
same here definition in the header ?

gtw2 replied:
Same as your previous comment, they are set with the vector now which is much clearer

katyhuff commented:
@gtw2 : What's the status here?

gtw2 commented:
@bam241 I forgot to ping you with that last commit, but I split the unit tests into pyre_tests (for the cyclus facility tests) and process_tests (for the SepMaterial functions used in each subprocess).

I also fixed the coefficients that were poorly defined - they are now defined in the header file with the vector as suggested. 

Side note - Should the process_tests be in src/ with pyre_tests or does it belong in src/toolkit/pyroprocessing?

katyhuff commented:
I think it looks good and it's passing the tests. Thoughts, @bam241 ? I know you had one remaining comment.

bam241 commented:
Sorry it took me so long for the last review !

Merging !
